compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolEnv.java

```diff
@@ -181,8 +181,14 @@ public static SymbolEnv createResourceActionSymbolEnv(BLangInvokableNode node, S
     }
 
     public static SymbolEnv createArrowFunctionSymbolEnv(BLangArrowFunction node, SymbolEnv env) {
-        SymbolEnv symbolEnv = new SymbolEnv(node, new Scope(env.scope.owner));
-        symbolEnv.enclEnv = env;
+        Scope scope = new Scope(env.scope.owner);
+        env.scope.entries.entrySet().stream()
+                // skip the type narrowed symbols when taking the snapshot for closures.
+                .filter(entry -> (entry.getValue().symbol.tag & SymTag.VARIABLE) != SymTag.VARIABLE ||
+                        ((BVarSymbol) entry.getValue().symbol).originalSymbol == null)
+                .forEach(entry -> scope.entries.put(entry.getKey(), entry.getValue()));
```

Comment:
Should we try and extract this logic out to a separate method, since it seems to be used across multiple methods?

Reply:
Yes, we should do it. Let me check.